crates/sui-core/src/authority_server.rs

```diff
@@ -439,27 +440,46 @@ impl ValidatorService {
         }
 
         // 5) Execute the certificate.
-        let span = tracing::debug_span!(
-            "validator_state_process_cert",
-            ?tx_digest,
-            tx_kind = certificate.signed_data.data.kind_as_str()
-        );
-
-        match state
-            .handle_certificate(&certificate)
-            .instrument(span)
-            .await
-            .map_err(|e| tonic::Status::internal(e.to_string()))
-        {
-            Err(e) => {
-                // Record the cert for later execution, including causal completion if necessary.
-                let tx_digest = *tx_digest;
-                let _ = state
-                    .add_pending_certificates(vec![(tx_digest, Some(certificate))])
-                    .tap_err(|e| error!(?tx_digest, "add_pending_certificates failed: {}", e));
-                Err(e)
+        // Often we cannot execute a cert due to dependenties haven't been executed, and we will
+        // observe ObjectErrors. In such case, we can wait and retry. It should eventually
+        // succeed.
+        // TODO: This is a quick hack. We should properly fix this through dependency-based
+        // scheduling.
+        let mut retry_cnt = 0;
+        loop {
+            let span = tracing::debug_span!(
+                "validator_state_process_cert",
+                ?tx_digest,
+                tx_kind = certificate.signed_data.data.kind_as_str()
+            );
+            match state
+                .handle_certificate(&certificate)
+                .instrument(span)
+                .await
+            {
+                err @ Err(SuiError::ObjectErrors { .. }) => {
+                    if retry_cnt > 500 {
```

Comment:
even for a temporary hack this will be aggressive when the shared obj is really hot:
1. the interval (100ms) is too short
2. overall time spend would be > 50s which is larger than the 30s default timeout so client side already times out

we should use a larger interval or use exponential backoff

Reply:
100ms not that short actually. A typical execution should take less than 100ms to finish, and that's all the next transaction needs to wait before it can succeed.
exponential backoff might be a bit problematic. Imagine the case where 10 transactions are all in-flight touching the same shared object:
The second and after will all fail when they first attempt execution. This would mean that each would only be able to wake up and succeed the next power of 2. At the end of the day, in this hack scheme, the allowed number of retries basically roughly determine the number of consecutive shared object tx we can support.

Reply:
This isn't quite right because execution driver will also be making attempts in the back ground

Reply:
I guess the case you mentioned can be solve by adding `jitter`, but mark's suggestion below `(desired_version - current_version) * 150ms` is better